core/drivers/clk/clk.c

```diff
@@ -296,3 +296,38 @@ TEE_Result clk_set_parent(struct clk *clk, struct clk *parent)
 
 	return res;
 }
+
+static TEE_Result clk_dummy_get_rates(struct clk *clk, size_t start_index,
+				      unsigned long *rates, size_t *nb_elts)
+{
+	if (start_index)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (!rates) {
+		*nb_elts = 1;
+		return TEE_SUCCESS;
+	}
+
+	if (*nb_elts != 1)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	rates[0] = clk_get_rate(clk);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result clk_get_rates_array(struct clk *clk, size_t start_index,
+			       unsigned long *rates, size_t *nb_elts)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	/* Simply return the clock rate */
+	if (!clk->ops->get_rates_array)
+		return clk_dummy_get_rates(clk, start_index, rates, nb_elts);
```

Comment:
If driver does not provide a get_rate_array handler, i think this function should return an error. If not, this function may report different rate if clock rate changes, which is not the desired behavior of this function.

Reply:
Acked. But even if moved to SCMI clock handling, I'm not sure we can easily play well with rate changes.

Reply:
If a clock driver does not register a `get_rates_array` operator, SCMI server will not expose possible rates and the SCMI client/agent won't request rate changes. Therefore, supporting rate changes mandates clock driver to describe all possible rates for that clock, whatever its parent is. Note that if switching to a new rate requires the clock to change its parent, it will be transparent from the SCMI client/agent view.

All in one, I think supporting exposure of the clock through the SCMI clock protocol should mandate the driver to register a `get_rates_array` operator.

Reply:
Ok, I was inspired by what was done for the stm32mp1 but I guess this was kind of a quirk. I will do as you told.
ie, get_rates_array *should* return the list of rates supported whatever the parent.